crates/pallet-rewards/src/tests.rs

```diff
@@ -0,0 +1,111 @@
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 400, 100, 99), 250);
+    assert_eq!(Pallet::update_avg_blockspace_usage(0, 400, 100, 99), 200);
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 0, 100, 99), 50);
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 400, 100, 100), 250);
+    assert_eq!(Pallet::update_avg_blockspace_usage(0, 400, 100, 100), 200);
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 0, 100, 100), 50);
+
+    // For `block_height` > `avg_blockspace_usage_num_blocks` we use the correct formula
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 400, 100, 101), 394);
+    assert_eq!(Pallet::update_avg_blockspace_usage(0, 400, 100, 101), 392);
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 0, 100, 101), 1);
+
+    // Special case check for no averaging blocks
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 400, 0, 100), 100);
+    assert_eq!(Pallet::update_avg_blockspace_usage(0, 400, 0, 100), 0);
+    assert_eq!(Pallet::update_avg_blockspace_usage(100, 0, 0, 100), 100);
+}
+
+#[test]
+fn correct_block_reward() {
+    // No reward unless parameters are set
+    assert_eq!(Pallet::block_reward(None, 0, 0), 0);
+
+    {
+        let params = IssuanceComponentParams {
+            initial_subsidy: SSC,
+            max_component_issuance: 1_000_000 * SSC,
+            initial_subsidy_start_block: 10,
+            initial_subsidy_duration: 10,
+        };
+        // No reward before initial subsidy start block
+        assert_eq!(Pallet::block_reward(Some(params), 9, 0), 0);
+        // Rewards starts at initial subsidy start block
+        assert_ne!(Pallet::block_reward(Some(params), 10, 0), 0);
+        // And remains the same for the whole initial subsidy duration
+        for offset in 0..params.initial_subsidy_duration {
+            assert_eq!(
+                Pallet::block_reward(Some(params), 10, 0),
+                Pallet::block_reward(Some(params), 10 + offset, 0)
+            );
+        }
+        // Blockspace usage (and storage fees as the result) mean lower block reward
+        assert!(
+            Pallet::block_reward(Some(params), 10, 1000)
+                < Pallet::block_reward(Some(params), 10, 0)
+        );
+        // Block reward goes down once initial subsidy ends
+        assert!(
+            Pallet::block_reward(Some(params), 10, 0) > Pallet::block_reward(Some(params), 21, 0)
+        );
+        assert!(
+            Pallet::block_reward(Some(params), 21, 0) > Pallet::block_reward(Some(params), 50, 0)
+        );
+        // Blockspace usage (and storage fees as the result) still mean lower block reward during delay period
+        assert!(
+            Pallet::block_reward(Some(params), 21, 100) < Pallet::block_reward(Some(params), 21, 0)
+        );
+        assert!(
+            Pallet::block_reward(Some(params), 50, 100) < Pallet::block_reward(Some(params), 50, 0)
+        );
+    }
```

Comment:
```suggestion
        //Taylor series approximation underflows at a particular block
        assert_ne!(Pallet::vote_reward(Some(params), 2_180_605), 0);
        assert_eq!(Pallet::vote_reward(Some(params), 2_180_606), 0);
    }
```
A test to see that we don't get to 0 rewards too fast.

Reply:
2M blocks is way too soon. I understand that we have 1M issuance in this test, but it still feels too soon. Maybe experiment with how many terms we need to add to Taylor series to extend it meaningfully just like the chart you have in the spec, but for more components?

Reply:
If we add terms up to 7th power, we can get to 2 758 995 (6.3 month total duration), 9th power to 3 333 538 (7.7 month total duration), 11th power to 3 905 432 (~9 month total duration), 13th power to 4 475 387 (10.2 month). Basically, every 2 powers add a month for this issuance.
On more realistic parameters, every 2 powers extend issuance by ~10 years. I totally recommend we use at least 7 or 9 if it doesn't overflow. Is that too heavy on compute weight or can we afford it?

Reply:
It should be fine since it is just a few multiplications and they shouldn't take seconds, but we need to test it. Can you add another version in the spec with corresponding coefficients up to 9th power and I'll do local benches to see how costly it is in WASM?